File: sdk/typescript/src/x402.ts
import type { ChargeRequest } from "./types";

/**
 * Parse an x402 charge header string into a typed ChargeRequest.
 *
 * The x402 header format is comma-separated key=value pairs:
 *   `amount=1000, asset=USDC, recipient=0x..., network=base-sepolia, nonce=uuidv7`
 *
 * This is a vendor copy of the parser from services/settlement/src/x402/parse.ts
 * to keep the SDK dependency-free from NestJS services.
 */
export function parseX402Header(header: string): ChargeRequest {
  if (!header || header.trim().length === 0) {
    throw new AgentPaySdkError("x402_parse_error", "Empty charge header");
  }

  const fields: Record<string, string> = {};

  for (const pair of header.trim().split(",")) {
    const trimmed = pair.trim();
    if (trimmed.length === 0) continue;

    const eqIndex = trimmed.indexOf("=");
    if (eqIndex === -1) {
      throw new AgentPaySdkError("x402_parse_error", `Malformed key=value pair: "${trimmed}"`);
    }

    const key = trimmed.slice(0, eqIndex).trim().toLowerCase();
    const value = trimmed.slice(eqIndex + 1).trim();

    if (key in fields) {
      throw new AgentPaySdkError("x402_parse_error", `Duplicate field: ${key}`);
    }

    fields[key] = value;
  }

  // Validate required fields
  for (const field of ["amount", "asset", "recipient", "network", "nonce"]) {
    if (!fields[field]) {
      throw new AgentPaySdkError("x402_parse_error", `Missing required field: ${field}`);
    }
  }

  if (fields.asset !== "USDC") {
    throw new AgentPaySdkError("x402_parse_error", `Unsupported asset: ${fields.asset}`);
  }

  if (!["base-mainnet", "base-sepolia"].includes(fields.network!)) {
    throw new AgentPaySdkError("x402_parse_error", `Unsupported network: ${fields.network}`);
  }

  if (!/^\d+$/.test(fields.amount!)) {
    throw new AgentPaySdkError("x402_parse_error", `Invalid amount: ${fields.amount}`);
  }

  if (!/^0x[0-9a-fA-F]{40}$/.test(fields.recipient!)) {
    throw new AgentPaySdkError("x402_parse_error", `Invalid recipient: ${fields.recipient}`);
  }

  return {
    amount: fields.amount!,
    asset: "USDC",
    recipient: fields.recipient as `0x${string}`,
    network: fields.network as "base-mainnet" | "base-sepolia",
    nonce: fields.nonce!,
  };
}

/**
 * SDK-specific error class.
 */
export class AgentPaySdkError extends Error {
  constructor(
    public readonly code: string,
    message: string,
  ) {
    super(message);
    this.name = "AgentPaySdkError";
  }
}

File: sdk/python/src/agentpay_sdk/x402.py
"""x402 charge header parser (mirrors sdk/typescript/src/x402.ts)."""

from .types import ChargeRequest, Network


REQUIRED_FIELDS = ("amount", "asset", "recipient", "network", "nonce")
VALID_NETWORKS = frozenset(["base-mainnet", "base-sepolia"])


class AgentPaySdkError(Exception):
    """SDK-specific error."""

    def __init__(self, code: str, message: str):
        self.code = code
        super().__init__(message)


def parse_x402_header(header: str) -> ChargeRequest:
    """Parse an x402 charge header string into a typed ChargeRequest.

    Format: ``amount=1000, asset=USDC, recipient=0x..., network=base-sepolia, nonce=uuidv7``
    """
    if not header or not header.strip():
        raise AgentPaySdkError("x402_parse_error", "Empty charge header")

    fields: dict[str, str] = {}

    for pair in header.strip().split(","):
        trimmed = pair.strip()
        if not trimmed:
            continue

        if "=" not in trimmed:
            raise AgentPaySdkError("x402_parse_error", f'Malformed key=value pair: "{trimmed}"')

        eq_index = trimmed.index("=")
        key = trimmed[:eq_index].strip().lower()
        value = trimmed[eq_index + 1 :].strip()

        if key in fields:
            raise AgentPaySdkError("x402_parse_error", f"Duplicate field: {key}")

        fields[key] = value

    # Validate required fields
    for field_name in REQUIRED_FIELDS:
        if field_name not in fields:
            raise AgentPaySdkError("x402_parse_error", f"Missing required field: {field_name}")

    if fields["asset"] != "USDC":
        raise AgentPaySdkError("x402_parse_error", f"Unsupported asset: {fields['asset']}")

    if fields["network"] not in VALID_NETWORKS:
        raise AgentPaySdkError("x402_parse_error", f"Unsupported network: {fields['network']}")

    if not fields["amount"].isdigit():
        raise AgentPaySdkError("x402_parse_error", f"Invalid amount: {fields['amount']}")

    if not _is_hex_address(fields["recipient"]):
        raise AgentPaySdkError("x402_parse_error", f"Invalid recipient: {fields['recipient']}")

    return ChargeRequest(
        amount=fields["amount"],
        asset="USDC",
        recipient=fields["recipient"],
        network=fields["network"],  # type: ignore[arg-type]
        nonce=fields["nonce"],
    )


def _is_hex_address(s: str) -> bool:
    """Check if a string is a valid 0x-prefixed 40-char hex address."""
    if not s.startswith("0x") or len(s) != 42:
        return False
    try:
        int(s[2:], 16)
        return True
    except ValueError:
        return False
